internal/rust/semantic/src/types.rs

```diff
@@ -0,0 +1,666 @@
+        self.flatten() == other.flatten()
+    }
+}
+
+impl Row {
+    // Records are implemented as a sequence or list of extenstions.
+    // This function extends a record by adding a new property to the
+    // head of the list.
+    fn extend(self, head: Property) -> Self {
+        Row::Extension {
+            head: head,
+            tail: Box::new(self),
+        }
+    }
+
+    // Flatten a record type into a hashmap of property names and types
+    fn flatten(&self) -> (HashMap<String, &MonoType>, Option<Tvar>) {
+        match self {
+            Row::Nil => (HashMap::new(), None),
+            Row::Var(tv) => (HashMap::new(), Some(*tv)),
+            Row::Extension { head: h, tail: t } => {
+                let (mut r, tv) = t.flatten();
+                r.insert(h.k.clone(), &h.v);
+                (r, tv)
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Property {
+    pub k: String,
+    pub v: MonoType,
+}
+
+impl fmt::Display for Property {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}:{}", self.k, self.v)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Function {
+    pub req: HashMap<String, MonoType>,
+    pub opt: HashMap<String, MonoType>,
+    pub pipe: Option<Property>,
+    pub retn: Option<MonoType>,
+}
+
+impl fmt::Display for Function {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut req: Vec<Property> = Vec::new();
+        for (k, v) in &self.req {
+            req.push(Property {
+                k: k.clone(),
+                v: v.clone(),
+            });
+        }
+        req.sort_unstable_by(|a, b| a.k.cmp(&b.k));
+
+        let mut opt: Vec<Property> = Vec::new();
+        for (k, v) in &self.opt {
+            opt.push(Property {
+                k: String::from("?") + &k,
+                v: v.clone(),
+            });
+        }
+        opt.sort_unstable_by(|a, b| a.k.cmp(&b.k));
+
+        let mut args: Vec<Property> = Vec::new();
+        if let Some(pipe) = &self.pipe {
+            if pipe.k == "<-" {
+                args.push(pipe.clone());
+            } else {
+                args.push(Property {
+                    k: String::from("<-") + &pipe.k,
+                    v: pipe.v.clone(),
+                });
+            }
+        }
+
+        args.append(&mut req);
+        args.append(&mut opt);
+        match &self.retn {
+            Some(t) => write!(f, "({}) -> {}", List { values: &args }, t),
+            None => write!(f, "({}) -> !", List { values: &args }),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+struct BoundTvar {
+    tv: Tvar,
+    kind: Kind,
+}
+
+impl fmt::Display for BoundTvar {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}:{}", self.tv, self.kind)
+    }
+}
+
+// List is a wrapper around a generic vector
+#[derive(Debug, Clone, PartialEq)]
+struct List<'a, T> {
+    values: &'a Vec<T>,
+}
```

Comment:
This is an internal type whose only purpose is for displaying a comma separated list of values.